base/compiler/optimize.jl

```diff
@@ -534,6 +534,26 @@ function ipo_escape_cache(mi_cache::MICache) where MICache
 end
 null_escape_cache(linfo::Union{InferenceResult,MethodInstance}) = nothing
 
+"""
+    run_minimal_passes(
+        ci::CodeInfo,
+        sv::OptimizationState,
+        caller::InferenceResult,
+    ) -> ir::IRCode
+
+Transform a `CodeInfo` to an `IRCode`.  Like [`run_passes`](@ref) but it does not run the
+majority of optimization passes.
+"""
+function run_minimal_passes(ci::CodeInfo, sv::OptimizationState, caller::InferenceResult)
+    @timeit "convert"   ir = convert_to_ircode(ci, sv)
+    @timeit "slot2reg"  ir = slot2reg(ir, ci, sv)
```

Comment:
I think it would be very rare to deal with `IRCode` without slog2reg (unless you are changing slog2reg itself). For one thing, `verify_ir` considers `IRCode` with slots invalid:

https://github.com/JuliaLang/julia/blob/2159bfba2f551bb22a92c6895ff3e46aa6dc5421/base/compiler/ssair/verify.jl#L47-L48

So, I think running slot2reg with `optimize = false` makes sense, if we treat it as "least optimized typical `IRCode` that compiler passes deal with." But running slot2reg manually is not super hard. So I don't have super strong opinion about it.

@aviatesk @ianatol Any thoughts?

Reply:
I guess we can also pass numeric `optimize_level`

```julia
function run_passes(ci::CodeInfo, sv::OptimizationState, caller::InferenceResult, optimize_level = typeinf(Int))
    stage = 1
    isdone() = optimize_level < (stage += 1)
    @timeit "convert"   ir = convert_to_ircode(ci, sv)
    isdone() && return ir
    @timeit "slot2reg"  ir = slot2reg(ir, ci, sv)
    isdone() && return ir
    # TODO: Domsorting can produce an updated domtree - no need to recompute here
    @timeit "compact 1" ir = compact!(ir)
    isdone() && return ir
    @timeit "Inlining"  ir = ssa_inlining_pass!(ir, ir.linetable, sv.inlining, ci.propagate_inbounds)
    isdone() && return ir
    ...
```

Reply:
I prefer running `slot2reg` as we often work on post-`slot2reg` `IRCode` when developing Julia-level optimizer. I also think the other "pre-inlining optimizations (mostly DCE)" happening within `convert_to_ircode` and `type_annotate!` is also significant anyway.

And the `optimize_level` idea sounds very useful!

Reply:
I added `optimize_level` 04797d59d3af59bb7db05a600b61151059fcaad0

```julia
julia> function demo(f)
           y = f()
           y isa Any ? nothing : f()
       end;

julia> Base.code_ircode(demo; optimize_level = 1) |> only |> first
2 1 ─      (_3 = (_2)())::#UNDEF                                                                   │
3 │   %2 = (_3 isa Main.Any)::#UNDEF                                                               │
  └──      goto #5 if not %2::#UNDEF                                                               │
  2 ─      return Main.nothing::#UNDEF                                                             │
  3 ─      Core.Const(:((_2)()))::#UNDEF                                                           │
  │        Core.Const(:(return %5))::#UNDEF                                                        │
  └──      unreachable::#UNDEF                                                                     │


julia> Base.code_ircode(demo; optimize_level = 2) |> only |> first
2 1 ─ %1 = (_2)()::Any                                                                             │
3 │   %2 = (%1 isa Main.Any)::Core.Const(true)                                                     │
  └──      goto #3 if not %2                                                                       │
  2 ─      return Main.nothing                                                                     │
  3 ─      Core.Const(:((_2)()))::Union{}                                                          │
  │        Core.Const(:(return %5))::Union{}                                                       │
  └──      unreachable                                                                             │


julia> Base.code_ircode(demo) |> only |> first
2 1 ─     (_2)()::Any                                                                              │
3 └──     goto #3 if not true                                                                      │
  2 ─     return Main.nothing                                                                      │
  3 ─     unreachable                                                                              │

```